js/framework/AnalyticsFramework.js

```diff
@@ -485,15 +485,61 @@ OO.Analytics.Framework = function()
     return success;
   };
 
-  //PublishMessage should only publish messages that are in OO.Analytics.EVENTS.
-  //To avoid doing an expensive string search through OO.Analyitcs.EVENTS for this checking,
-  //here we convert OO.Analytics.EVENTS into another object where we can directly look up if
-  //the message exists.
-  for(var tempEventName in OO.Analytics.EVENTS)
+  /**
+   * Helper function to flatten an object with a nested objects into a single array of values.
```

Comment:
This function didn't flatten the data into a string.  It took all the constants in OO.Analytics.Events and made an object lookup table.  So it could do constant lookup.  Creating a list means linear search everytime.